// genezio/build/src/models/cloudProviderIdentifier.js
/*
 * Use cloud adapter to alternate between the supported cloud infrastructures.
 * This information is not meant to be user-facing.
 * It's meant to organize resources on genezio infrastructure.
 */
export var CloudAdapterIdentifier;
(function (CloudAdapterIdentifier) {
    CloudAdapterIdentifier["AWS"] = "aws";
    CloudAdapterIdentifier["RUNTIME"] = "runtime";
    CloudAdapterIdentifier["CLUSTER"] = "cluster";
})(CloudAdapterIdentifier || (CloudAdapterIdentifier = {}));
export var CloudProviderIdentifier;
(function (CloudProviderIdentifier) {
    CloudProviderIdentifier["GENEZIO_CLOUD"] = "genezio-cloud";
    CloudProviderIdentifier["GENEZIO_UNIKERNEL"] = "genezio-unikernel";
    CloudProviderIdentifier["GENEZIO_CLUSTER"] = "genezio-cluster";
    CloudProviderIdentifier["GENEZIO_AWS"] = "genezio-aws";
    CloudProviderIdentifier["GENEZIO_LEGACY"] = "genezio";
    CloudProviderIdentifier["CAPYBARA_LEGACY"] = "capybara";
    CloudProviderIdentifier["CAPYBARA_LINUX_LEGACY"] = "capybaraLinux";
    CloudProviderIdentifier["CLUSTER_LEGACY"] = "cluster";
})(CloudProviderIdentifier || (CloudProviderIdentifier = {}));
export const CloudProviderMapping = {
    [CloudProviderIdentifier.CAPYBARA_LEGACY]: CloudProviderIdentifier.GENEZIO_UNIKERNEL,
    [CloudProviderIdentifier.CAPYBARA_LINUX_LEGACY]: CloudProviderIdentifier.GENEZIO_CLOUD,
    [CloudProviderIdentifier.GENEZIO_LEGACY]: CloudProviderIdentifier.GENEZIO_AWS,
    [CloudProviderIdentifier.CLUSTER_LEGACY]: CloudProviderIdentifier.GENEZIO_CLUSTER,
};
