// genezio/build/src/models/genezioModels.js
export var GenezioCommandTemplates;
(function (GenezioCommandTemplates) {
    GenezioCommandTemplates["FULLSTACK"] = "Fullstack";
    GenezioCommandTemplates["BACKEND"] = "Backend-Only";
})(GenezioCommandTemplates || (GenezioCommandTemplates = {}));
export class File {
    constructor(path, data) {
        this.path = path;
        this.data = data;
    }
}
export class SdkFileClass {
    constructor(path, data, className) {
        this.path = path;
        this.data = data;
        this.className = className;
    }
}
export var AstNodeType;
(function (AstNodeType) {
    AstNodeType["StringLiteral"] = "StringLiteral";
    AstNodeType["IntegerLiteral"] = "IntegerLiteral";
    AstNodeType["BooleanLiteral"] = "BooleanLiteral";
    AstNodeType["FloatLiteral"] = "FloatLiteral";
    AstNodeType["NullLiteral"] = "NullLiteral";
    AstNodeType["DoubleLiteral"] = "DoubleLiteral";
    AstNodeType["BigIntLiteral"] = "BigIntLiteral";
    AstNodeType["VoidLiteral"] = "VoidLiteral";
    AstNodeType["AnyLiteral"] = "AnyLiteral";
    AstNodeType["ArrayType"] = "ArrayType";
    AstNodeType["DateType"] = "DateType";
    AstNodeType["MapType"] = "MapType";
    AstNodeType["PromiseType"] = "PromiseType";
    AstNodeType["ConstType"] = "ConstType";
    AstNodeType["NativeType"] = "NativeType";
    AstNodeType["ParamType"] = "ParamType";
    AstNodeType["CustomNodeLiteral"] = "CustomNodeLiteral";
    AstNodeType["Enum"] = "Enum";
    AstNodeType["TypeAlias"] = "TypeAlias";
    AstNodeType["TypeLiteral"] = "TypeLiteral";
    AstNodeType["StructLiteral"] = "StructLiteral";
    AstNodeType["UnionType"] = "UnionType";
    AstNodeType["ParameterDefinition"] = "ParameterDefinition";
    AstNodeType["MethodDefinition"] = "MethodDefinition";
    AstNodeType["ClassDefinition"] = "ClassDefinition";
    AstNodeType["PropertyDefinition"] = "PropertyDefinition";
})(AstNodeType || (AstNodeType = {}));
export var SourceType;
(function (SourceType) {
    SourceType["script"] = "script";
    SourceType["module"] = "module";
})(SourceType || (SourceType = {}));
export var MethodKindEnum;
(function (MethodKindEnum) {
    MethodKindEnum["constructor"] = "constructor";
    MethodKindEnum["method"] = "method";
    MethodKindEnum["get"] = "get";
    MethodKindEnum["set"] = "set";
})(MethodKindEnum || (MethodKindEnum = {}));
