// genezio/build/src/models/projectOptions.js
export const DEFAULT_NODE_RUNTIME = "nodejs20.x";
export const DEFAULT_ARCHITECTURE = "x86_64";
export const DEFAULT_PYTHON_RUNTIME = "python3.13.x";
export const DEFAULT_PYTHON_VERSION_INSTALL = "3.13";
export const CONTAINER_IMAGE_NODE20 = "node:20.11.1-alpine3.19";
export const DEFAULT_NODE_RUNTIME_IMAGE = CONTAINER_IMAGE_NODE20;
// Note: ts and tsx are not included in the list of FUNCTION_EXTENSIONS intentionally
// A typescript function will be compiled to javascript before being deployed
export const FUNCTION_EXTENSIONS = ["js", "mjs", "cjs", "py"];
export var SSRFrameworkComponentType;
(function (SSRFrameworkComponentType) {
    SSRFrameworkComponentType["next"] = "nextjs";
    SSRFrameworkComponentType["nitro"] = "nitro";
    SSRFrameworkComponentType["nuxt"] = "nuxt";
    SSRFrameworkComponentType["nestjs"] = "nestjs";
    SSRFrameworkComponentType["remix"] = "remix";
    SSRFrameworkComponentType["streamlit"] = "streamlit";
})(SSRFrameworkComponentType || (SSRFrameworkComponentType = {}));
// These are the human friendly names for the SSR frameworks
// They can be used for displaying prettify logs
export const SSRFrameworkName = {
    [SSRFrameworkComponentType.next]: "Next.js",
    [SSRFrameworkComponentType.nitro]: "Nitro",
    [SSRFrameworkComponentType.nuxt]: "Nuxt.js",
    [SSRFrameworkComponentType.nestjs]: "NestJS",
    [SSRFrameworkComponentType.remix]: "Remix",
    [SSRFrameworkComponentType.streamlit]: "Streamlit",
};
export var ContainerComponentType;
(function (ContainerComponentType) {
    ContainerComponentType["container"] = "container";
})(ContainerComponentType || (ContainerComponentType = {}));
export const supportedNodeRuntimes = ["nodejs20.x"];
export const supportedPythonRuntimes = [
    "python3.9.x",
    "python3.10.x",
    "python3.11.x",
    "python3.12.x",
    "python3.13.x",
];
export const supportedArchitectures = ["arm64", "x86_64"];
export const supportedSSRFrameworks = ["nextjs", "nitro", "nuxt", "streamlit"];
export const supportedPythonDepsInstallVersion = ["3.9", "3.10", "3.11", "3.12", "3.13"];
