// genezio/build/src/requests/authentication.js
import sendRequest from "../utils/requests.js";
export async function setAuthentication(envId, request) {
    const data = JSON.stringify(request);
    const response = (await sendRequest("PUT", `core/auth/${envId}`, data));
    return response;
}
export async function getAuthentication(envId) {
    const response = (await sendRequest("GET", `core/auth/${envId}`, ""));
    return response;
}
export async function getAuthProviders(envId) {
    const response = (await sendRequest("GET", `core/auth/providers/${envId}`, ""));
    return response;
}
export async function setAuthProviders(envId, request) {
    const data = JSON.stringify(request);
    const response = (await sendRequest("PUT", `core/auth/providers/${envId}`, data));
    return response;
}
export async function getEmailTemplates(envId) {
    const response = (await sendRequest("GET", `core/auth/email-templates/${envId}`, ""));
    return response;
}
export async function setEmailTemplates(envId, request) {
    const data = JSON.stringify(request);
    const response = (await sendRequest("PUT", `core/auth/email-templates/${envId}`, data));
    return response;
}
