// genezio/build/src/requests/getPresignedURLForProjectCodePush.js
import axios from "./axios.js";
import { getAuthToken } from "../utils/accounts.js";
import { BACKEND_ENDPOINT } from "../constants.js";
import version from "../utils/version.js";
import { GENEZIO_NOT_AUTH_ERROR_MSG, UserError } from "../errors.js";
export async function getPresignedURLForProjectCodePush(region = "us-east-1", projectName, stage) {
    if (!region || !projectName) {
        throw new UserError("Missing required parameters");
    }
    // Check if user is authenticated
    const authToken = await getAuthToken();
    if (!authToken) {
        throw new UserError(GENEZIO_NOT_AUTH_ERROR_MSG);
    }
    const json = JSON.stringify({
        projectName: projectName,
        region: region,
        stage: stage,
    });
    const response = await axios({
        method: "POST",
        url: `${BACKEND_ENDPOINT}/core/create-project-code-url`,
        data: json,
        headers: {
            Authorization: `Bearer ${authToken}`,
            "Accept-Version": `genezio-cli/${version}`,
        },
        maxContentLength: Infinity,
        maxBodyLength: Infinity,
    });
    return response.data.presignedURL;
}
