// genezio/build/src/requests/integration.js
import sendRequest from "../utils/requests.js";
export async function enableEmailIntegration(projectId, projectEnvId) {
    return enableIntegration({
        integrationName: "EMAIL-SERVICE",
    }, projectId, projectEnvId);
}
export async function disableEmailIntegration(projectId, projectEnvId) {
    return disableIntegration({
        integrationName: "EMAIL-SERVICE",
    }, projectId, projectEnvId);
}
export async function enableIntegration(request, projectId, projectEnvId) {
    const { integrationName, envVars = [] } = request;
    const data = JSON.stringify({
        integrationName: integrationName,
        envVars: envVars,
    });
    const status = (await sendRequest("POST", `projects/${projectId}/${projectEnvId}/integrations`, data));
    return status;
}
export async function disableIntegration(request, projectId, projectEnvId) {
    const { integrationName, envVars = [] } = request;
    const data = JSON.stringify({
        integrationName: integrationName,
        envVars: envVars,
    });
    const status = (await sendRequest("DELETE", `projects/${projectId}/${projectEnvId}/integrations`, data));
    return status;
}
export async function getProjectIntegrations(projectId, projectEnvId) {
    const response = (await sendRequest("GET", `projects/${projectId}/${projectEnvId}/integrations`, ""));
    return response;
}
